GitHub analysis functions

taoTimTim opened the pull request:
<!-- 
Thank you for contributing! Please fill out this template to help us review your PR.
-->

## 📝 Description

This PR introduces the initial modules for determining a user's collaboration skills.
We retrieve GitHub project metrics using the GitHub REST API, and once those metrics are available, they are passed into the functions implemented in this PR to compute the user’s collaboration skill profile. The flow is not implemented in this PR, only the functions and models required are.

This PR includes:
- All modules and functions required for the initial skill computation flow
- A new CollaborationProfile model to store the computed skill results
- Supporting computation classes
- Full unit test coverage for the new functionality

This is not the full implementation; this PR only includes the foundation. Additional computation logic and end-to-end integration will be added in future PRs to keep this one a manageable size.

This PR continues the work started in the GitHub analysis pipeline.
**Works On:** #205 

---

## 🔧 Type of Change

- [ ] 🐛 Bug fix (non-breaking change that fixes an issue)
- [x] ✨ New feature (non-breaking change that adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 📚 Documentation added/updated
- [x] ✅ Test added/updated
- [ ] ♻️ Refactoring
- [ ] ⚡ Performance improvement

---

## 🧪 Testing

- Ran `pytest`, new and old tests are passing
- Ran the system with a code, collaborative project to ensure no errors were created when implementing these functions.

---

## ✓ Checklist

- [x] 🤖 GenAI was used in generating the code and I have performed a self-review of my own code
- [x] 💬 I have commented my code where needed
- [ ] 📖 I have made corresponding changes to the documentation
- [x] ⚠️ My changes generate no new warnings
- [x] ✅ I have added tests that prove my fix is effective or that my feature works and tests are passing locally
- [x] 🔗 Any dependent changes have been merged and published in downstream modules
- [ ] 📱 Any UI changes have been checked to work on desktop, tablet, and/or mobile

---

## 📸 Screenshots

<details>
<summary>Click to expand screenshots</summary>

No screenshots necessary for this, it is all backend.

</details>

ammaarkhan reviewed (approved):
Hey Timmi, excellent work on the GitHub collaboration profile foundation. The architecture is really clean. I like how you've separated each metric into its own module (normalized contributions, review quality, participation, etc.). 

A few things I noticed while reviewing:
Hardcoded weights: I see weights like 1.2 and 1.4 in leadership.py:9-10 and participation.py:17-21. Adding a small comment explaining the weighting logic would help us remember and understand why these numbers were initially chosen.
Review quality heuristics: The 40-character threshold and scoring formula in review_quality.py:19,23 are interesting choices. A quick comment justifying the approach would be helpful for anyone tweaking these later.

These are pretty minor documentation gaps that don't block anything. Ready to merge. All tests passing: 

<img width="1290" height="598" alt="Image" src="https://github.com/user-attachments/assets/12cd5f03-b859-40b6-80a5-e869e13c1605" />

taoTimTim commented:
@salmavkh Hey, thanks for the review. I incorporated your ideas in the detectors, if you could take a look at those and confirm that's wht you meant? I also clarified some of the comments and fixed the tests for the changes.

salmavkh reviewed (approved):
@taoTimTim thanks for making the changes! The word-diversity check in review_quality.py now filters out repetitive comments, and pr_review_weight in participation.py gives high-quality reviews some extra credit. Not sure yet if the unique-word minimum should be higher, but this seems fine for now.
